### PR TITLE
feat: Async search client

### PR DESCRIPTION
**Pull request by adilansari**
- User friendly search request builder. Introduced following capabilities:
```java
SearchRequest.newBuilder("my search").build();
SearchRequest.newBuilder("my search").withSearchFields("first_name", "last_name").build();
SearchRequest.newBuilder("my search").withFacetFields("first_name", "last_name").build();
```

- Updated readme docs
- Async search client

### Checks
- [x] Unit tests
- [x] Javadocs publishing passes without error
- [x] Coverage for new features

### CODE REVIEW

File: client/src/main/java/com/tigrisdata/db/client/TigrisAsyncCallback.java [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+ */
+
+package com.tigrisdata.db.client;
+
+/**
+ * Asynchronous callback to receive messages from server
+ *
+ * @param <T> type of message
+ */
+public interface TigrisAsyncCallback<T> {
```
Comment (adilansari) [inline client/src/main/java/com/tigrisdata/db/client/TigrisAsyncCallback.java:22]: Interface for TigrisAsyncReader, TigrisAsyncSearchReader and TigrisAsyncStreamEvent

File: client/src/test/java/com/tigrisdata/db/client/collection/User.java [hunk 2]
```diff
@@ -17,13 +17,16 @@
 
 /** Test collection model */
 public class User implements TigrisCollectionType {
-  String name;
   int id;
+  String name;
+  double salary;
```
Comment (adilansari) [inline client/src/test/java/com/tigrisdata/db/client/collection/User.java:22]: This is not used anywhere except `README` examples.

File: client/src/test/java/com/tigrisdata/db/client/StandardTigrisAsyncCollectionFailureTest.java [hunk 3]
```diff
@@ -152,4 +159,40 @@ public void onCompleted() {
...
+    while (!completed.get() && timeout < 20) {
+      timeout++;
+      try {
+        //noinspection BusyWait
+        Thread.sleep(100);
+      } catch (InterruptedException ignore) {
+      }
+    }
+
+    // there must not be any errors
```
Comment (JigarJoshi) [inline client/src/test/java/com/tigrisdata/db/client/StandardTigrisAsyncCollectionFailureTest.java:195]: comment is contradictory of assertion.

File: client/src/test/java/com/tigrisdata/db/client/search/SearchRequestTest.java [hunk 4]
```diff
@@ -48,8 +48,24 @@ public void build() {
...
+    Assert.assertEquals(expectedQuery, actual.getQuery());
+    Assert.assertEquals(expectedSearchFields, actual.getSearchFields());
+    Assert.assertEquals(expectedFacetQuery, actual.getFacetQuery());
+  }
+
   @Test
   public void failsWithNullQuery() {
-    Assert.assertThrows(NullPointerException.class, () -> SearchRequest.newBuilder(null).build());
+    Assert.assertThrows(
+        NullPointerException.class, () -> SearchRequest.newBuilder((String) null).build());
```
Comment (JigarJoshi) [inline client/src/test/java/com/tigrisdata/db/client/search/SearchRequestTest.java:69]: let's wrap it to IllegalArgumentException with proper message on what is not expected to be null.

File: README.md [hunk 5]
```diff
@@ -75,6 +75,13 @@ users.delete(
         Filters.eq("id", 2)
     )
 );
+
+// search - search for users with name "Jania"
+users.search(
+    SearchRequest.newBuilder("Jania")
+      .withSearchFields("name")
+      .build()
+    );
```
Comment (JigarJoshi) [inline README.md:84]: looks much cleaner now.

**Review by JigarJoshi — APPROVED** [review]

**Pull request merged**